### PR TITLE
expression: rewrite builtin function: FIND_IN_SET

### PR DESCRIPTION
**Pull request by zz-jason**
to #4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1999,49 +1999,46 @@ func (c *findInSetFunctionClass) getFunction(args []Expression, ctx context.Cont
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinFindInSetSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = 3
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: why set 3 here?

Comment (zz-jason) [inline expression/builtin_string.go]: copy from mysq: "Item_func_find_in_set::fix_length_and_dec()"

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1999,49 +1999,46 @@ func (c *findInSetFunctionClass) getFunction(args []Expression, ctx context.Cont
...
-	if len(strlst) == 0 {
-		return
+	if len(strlist) == 0 {
+		return 0, false, nil
 	}
-	for i, s := range strings.Split(strlst, ",") {
-		if s == str {
-			d.SetInt64(int64(i + 1))
-			return
+
```
Comment (winkyao) [inline expression/builtin_string.go]: > This function does not work properly if the first argument contains a comma (,) character.

File: expression/integration_test.go [hunk 3]
```diff
@@ -792,6 +792,12 @@ func (s *testIntegrationSuite) TestStringBuiltin(c *C) {
 	result.Check(testkit.Rows("0 1777777777777777777777 1777777777777777777777 1777777777777777777777"))
 	result = tk.MustQuery(`select oct(-1.9), oct(1.9), oct(-1), oct(1), oct(-9999999999999999999999999), oct(9999999999999999999999999);`)
 	result.Check(testkit.Rows("1777777777777777777777 1 1777777777777777777777 1 1777777777777777777777 1777777777777777777777"))
+
+	// for find_in_set
+	result = tk.MustQuery(`select find_in_set("", ""), find_in_set("", ","), find_in_set("中文", "字符串,中文");`)
+	result.Check(testkit.Rows("0 1 2"))
+	result = tk.MustQuery(`select find_in_set(NULL, ""), find_in_set("", NULL), find_in_set(1, "2,3,1");`)
+	result.Check(testkit.Rows("<nil> <nil> 3"))
```
Comment (winkyao) [inline expression/integration_test.go]: add `select find_in_set('b,', 'a,b,c,d')`

MySQL:
```
mysql> select find_in_set('b,', 'a,b,c,d');
+------------------------------+
| find_in_set('b,', 'a,b,c,d') |
+------------------------------+
|                            0 |
+------------------------------+
1 row in set (0.00 sec)
```

Comment (zz-jason) [inline expression/integration_test.go]: added

**Pull request merged**